Issue 3601 -- astropy.stats.funcs.bootstrap now accepts bootfuncs with multiple outputs

ezbc opened the pull request:
I have addressed [Issue 3601](https://github.com/astropy/astropy/issues/3601). I implemented the option for users to supply a function with multiple outputs to ``bootfunc``. They can control which ``bootfunc`` outputs to retain with ``output_index``.

This function could be sped up if the ``if`` statements were moved outside of the loop.

embray commented:
@ezbc Your PR so far has the changelog entry and some tests which look good, but is missing the actual change to the function.  Is it intentionally not implemented yet?

ezbc commented:
@embray This was not intentional, sorry about that. I must have committed funcs to a different branch. 

I believe the changes are now pushed to the correct branch, issue3601, and are now ready to be merged.

bsipocz commented on astropy/stats/tests/test_funcs.py:
```diff
@@ -9,6 +9,8 @@
 from numpy.testing import assert_equal
 from numpy.testing.utils import assert_allclose
 
+from scipy.stats import spearmanr
```
scipy is an optional dependency at the moment, so this import should be done in your test function, which should be decorated with ``@pytest.mark.skipif('not HAS_SCIPY')``. There are a few examples below of how to do that.

bsipocz commented on astropy/stats/tests/test_funcs.py:
```diff
@@ -236,6 +238,60 @@ def test_bootstrap():
+        answer = np.array((0.19425, 0.02094))
+
+        bootresult = funcs.bootstrap(bootarr, 2,
+                                     bootfunc=spearmanr)
```
I think you may want to put this new part in a separate function as it depends on scipy. The old test can run without scipy, and this function gets the decorator and runs only when scipy is available.

bsipocz commented on astropy/stats/funcs.py:
```diff
@@ -728,26 +735,65 @@ def bootstrap(data, bootnum=100, samples=None, bootfunc=None):
     # make sure the input is sane
     assert samples > 0, "samples cannot be less than one"
     assert bootnum > 0, "bootnum cannot be less than one"
+    assert type(data) is np.ndarray, "data must be a numpy.ndarray"
```
I believe using ``isinstance()`` is the recommended way to check whether the object type is the expected one.

bsipocz commented:
@ezbc - Thanks for the contribution. I left a few basic comments for an issue that causes test failures.

 Otherwise I'll leave the code review to others who are more familiar with this module.

eteq commented:
@ezbc @bsipocz (and @ttshimiz , who created #3601) - Sorry I'm only saying this now (I didn't think about it until I started reviewing this in detail), but I'm not sure this is the best solution to issue #3601 .  I think this approach adds unnecessary complication where a reasonably straightforward solution already exists in python: defining a new function.  For example, the following works the same as the case here that uses ``output_index=0``:
```
bootstrap(bootarr, 2, bootfunc=lambda x:spearmanr(x)[0])
```

I think this might be better because it makes the `bootstrap` function itself simpler while at the same time allowing other more complicated situations e.g. you want to use a different *input* (i.e., ``bootstrap(bootarr, bootfunc=lambda x: func(1, x))``.

So that would require no changes to the code... *but*, @ezbc, most of the work you've done here is adding tests, and we certainly don't want to waste that.  So perhaps the think to do is to back out the changes to the code, but then adapt the tests to use the lambda function approach, and *also* add an example to the documentation showing this use case?  @ttshimiz, would you consider that an acceptable solution to #3601 ?


(@ezbc - just in case, no need to worry that this will affect your GSoC application even if we make the changes - you'll still get credit for the effort you put in here as far as the application is concerned no matter how this PR actually turns out, I promise!)

ttshimiz commented:
@ezbc, @bsipocz, @eteq Yeah that would definitely be an acceptable solution to me. Examples and an explanation in the documentation would really help to show people how to use bootstrap with functions that have multiple inputs and outputs. It didn't even occur to me to use lambda to just define a new function that only returned the first output of spearmanr. In my view, simpler is always better. Thanks for addressing this issue!

ezbc commented:
Thanks for input @embray! Good thinking. I'll try to work in your suggestion into the documentation and testing this week.

embray commented:
@ezbc I think you meant to thank @eteq for his input.  Yes, there are two Eriks in this thread :)  It's easy to remember because the other one's username ends in q which is sort of odd :)

ezbc commented:
Aha, good suggestion @embray. 

Thanks for input @eteq! I'll work on the suggestion today.

ezbc commented:
@eteq, one question. I'm not sure when this case may arise, but perhaps someone would want to bootstrap data, retaining two values from the output of ``bootfunc``. I do not know how to do this using lambda functions besides running the ``bootstrap`` function twice with the same RNG seed, retaining a different output from the ``bootfunc``.

Can you think of a simpler way to retain two values from ``bootfunc``, or is this even necessary? If using a seed with two ``bootstrap`` runs is necessary, should we include an optional argument in the ``bootstrap`` function to accept a seed number? Otherwise, I suppose I could just include an example in the documentation of using ``astropy.utils.misc.NumpyRNGContext`` to supply a seed to ``bootfunc``.

Thanks!

eteq commented:
@ezbc - Ah, I hadn't thought of that.  Not sure how often it would be useful, but it might well be!

So this leads me to think the best thing is a "hybrid" solution: modify the function so that it still has no new inputs, but the *output* can be multi-dimensional.  So if bootfunc returns a single value, have it behave just as it does now, but if it's multiple values, then the output is an NxM array where N is the number of resamples and M is the number of return values that bootfunc yields.  I'd still advocate the lambda function example be included, but with that change, an alternative solution is ``bootstrap(bootarr, 2, bootfunc=spearmanr)[:, 0]``, which is even simpler.  This won't work if bootfunc gives back something irregular (i.e., if the first return value is an array and the second is a scalar), but then the situation is probably complex enough that the user should just do it themselves...

eteq commented:
@ezbc - do you think you'll be able to get to doing this?

ezbc commented:
@eteq - My apologies, lots on the plate in the past month, and now I'm going on holiday. I will work on this the last week of June if the issue is still open.

ezbc commented:
@eteq - I made the changes we discussed:
+ reverted the bootstrap function to the original, with the added ability for multiple outputs
+ added examples to the bootstrap function
+ changed testing to include single output and multiple output tests of bootfuncs

Sorry this took so long! Let me know if there's something I should change/add.

astrofrog commented:
@ezbc - could you address the test failures on Travis?

ezbc commented:
@astrofrog - I believe I addressed these issues. 

Would you be able to point me to a resource where I could learn to make these build checks locally myself, to avoid such issues in the future? Thanks!

astrofrog commented:
@ezbc - there is no easy way to run all the configurations, but you can run the tests with

```
python setup.py test
```

What I normally do is have a Python 2 and Python 3 environment in Anaconda, and I can run the tests in both of these. But no one really tests *all* the configurations before pushing a pull request. If you see Travis failures that are not obvious, you can use conda to set up a similar environment.

ezbc commented:
Thank you, I'll try to use Anaconda in this way in the future.

On Thu, Oct 8, 2015 at 4:18 AM, Thomas Robitaille <notifications@github.com>
wrote:

> @ezbc <https://github.com/ezbc> - there is no easy way to run all the
> configurations, but you can run the tests with
>
> python setup.py test
>
> What I normally do is have a Python 2 and Python 3 environment in
> Anaconda, and I can run the tests in both of these. But no one really tests
> *all* the configurations before pushing a pull request. If you see Travis
> failures that are not obvious, you can use conda to set up a similar
> environment.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/astropy/astropy/pull/3628#issuecomment-146467817>.
>



-- 
Elijah Bernstein-Cooper
ezbc@astro.wisc.edu
Department of Astronomy
University of Wisconsin - Madison

crawfordsm commented:
So all set for a final review?  I'm happy to give it one if that is the case.

ezbc commented:
I believe so. Let me know if there's anything which needs editing.

On Fri, Oct 16, 2015 at 5:06 AM, Steve Crawford <notifications@github.com>
wrote:

> So all set for a final review? I'm happy to give it one if that is the
> case.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/astropy/astropy/pull/3628#issuecomment-148673976>.
>



-- 
Elijah Bernstein-Cooper
ezbc@astro.wisc.edu
Department of Astronomy
University of Wisconsin - Madison

crawfordsm commented on astropy/stats/funcs.py:
```diff
@@ -725,7 +725,57 @@ def bootstrap(data, bootnum=100, samples=None, bootfunc=None):
-        Bootstrapped data. Each row is a bootstrap resample of the data.
+        Bootstrapped data. Each row is a bootstrap resample of the data. The
+        columns will correspond to the outputs of bootfunc.
+
```
Can this be clarified a little bit as this changes depending on whether or not bootfunc is specified.   So something like:  'If bootfunc is None, each row is a bootstrap resample of the data.   If bootfunc is specified, the columns of each row will correspond to the outputs of bootfunc'.

crawfordsm commented:
One relatively small edit and otherwise, this looks good to merge.  Thanks for the contribution @ezbc !

ezbc commented:
All done! Thank you all!

crawfordsm commented:
@embray this appears to be failing the coverage test but it doesn't appear to be complaining about any files that are in the commit.  Any idea why that is or what might causing the failure?   And if not, should this just be merged?

embray commented:
Whoever's in charge of this module should look at the coverage report and see what the coverage change is in any actually relevant modules.

crawfordsm commented:
That's me and I'm just confused as when I look at the coverage report, the one file that is changed as part of this PR has an increase in coverage, not a decrease.

bsipocz commented:
@crawfordsm @embray - It says that the PR base is [travis built No 10655](https://travis-ci.org/astropy/astropy/builds/83779569) but that't not the latest master. There are changes in vo and wcs since, so I guess that may somehow affect the result as those are the one with big coverage changes in this report?

bsipocz commented:
So my working hypothesis is if a PR is on top of an "old" master, and the current master changes that gradually decreased the coverage, and by the current master it's got below the threshold?

ezbc commented:
Sorry for the trouble. Should I merge issue3601 branch into master?

bsipocz commented:
@ezbc - No, I don't think you should because of this. This seems to be a totally unrelated bug/feature of coveralls.

However the changelog entry seems to be placed under the wrong section, I think it should go under 1.2 and not 0.3 (@crawfordsm is the boss, so he may point you to another bugfix release, but it won't be 0.3 in any case).

crawfordsm commented:
Good catch @bsipocz -- yes the changelog should current be under a 1.2 section and thanks for the help as well.  This is an update and not a bug fix. 

@ezbc -- please update this.  When you do, make sure to pull any changes upstream into your current branch as well and hopefully when you do another commit to this branch, it will build against the current master and not an older version.

ezbc commented:
I believe I merged the upstream changes into the current branch. Hopefully this worked!

bsipocz commented:
OK, it worked. 
However merging/pulling upstream is not the recommended way to update a feature branch [1]. So when you fix the CHANGES entry, could you please rebase it to remove the "merge" commit?

[1] http://docs.astropy.org/en/latest/development/workflow/development_workflow.html#astropy-guidelines-for-git

ezbc commented:
Sorry about that, it's my first time contributing. I'll follow the guidelines in the future.

Let me know if I missed something.

crawfordsm commented:
@astrofrog Just checking, it is okay to merge this as the failures are unrelated to the pull request?

astrofrog commented:
@crawfordsm - I've restarted AppVeyor

crawfordsm commented:
Merging-- thanks @ezbc for the contribution and everyone else for the review!

ezbc commented:
Thank you all for your patience!

On Mon, Nov 9, 2015 at 9:58 PM, Steve Crawford <notifications@github.com>
wrote:

> Merged #3628 <https://github.com/astropy/astropy/pull/3628>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/astropy/astropy/pull/3628#event-459639927>.
>



-- 
Elijah Bernstein-Cooper
ezbc@astro.wisc.edu
Department of Astronomy
University of Wisconsin - Madison

eteq commented:
@crawfordsm @ezbc - can someone clarify for me what the version is for this PR?  It looks to me like it came in *after* 1.1 feature freeze, which means it should be v1.2 .  But it has the v1.1 milestone and appears in the 1.1 changelog.  Is that an oversight?

I'm going to re-milestone it as 1.2 assuming that's the case (and there will be a reckoning later to make the `master` changelog consistent with what's actually in v1.1, but if I'm misunderstanding please let me know ASAP.

crawfordsm commented:
@eteq It did only get merged in after the 1.1 feature freeze so unfortunately due to various reasons it was not in time for 1.1.   I think I mention up earlier in the thread it the change log should have been moved to the 1.2 version.  So since this can't go into the 1.1, then it should be 1.2.   Let me know if there is something I can do to help with the change log as I must have missed that in the final review.

crawfordsm commented:
Although @embray looks to added the 1.1 milestone 8 days ago, so perhaps he should comment as well.

embray commented:
I don't really care, but the changelog entry was made for v1.1 so that's why I put it there.

eteq commented:
OK, thanks for clearing that up - I'll just move it to the right changelog section directly in master.
